src/Psy/Readline/TabCompletion.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Psy\Readline;
+
+use Psy\Context;
+
+/**
+ * Class TabCompletion
+ * @package Psy\Readline
+ */
+class TabCompletion
+{
+    /** @var Context */
+    protected $context;
+
+    public function __construct(Context $context = null)
+    {
+        $this->context = $context;
+    }
+
+    public function activate()
+    {
+        readline_completion_function(array(&$this, 'callback'));
+    }
+
+    /**
+     * @param $input
+     * @param $index
+     * @return array
+     */
+    public function callback($input, $index)
+    {
+        $info = readline_info();
+
+        // the char just before the cursor is a dollar? send a variable context
+        $charAt = substr($info['line_buffer'], $index-1, 1);
+        if ($charAt == '$') {
+            return array_keys($this->context->getAll());
+        }
+
+        $parenthesize = function ($name) {
+            return sprintf('%s()', $name);
```

Comment:
Yes, It could be done with reflection, but as you said, the objects might need to be cached because the reflected entities would be spending huge amounts of memory if they are not stored in a certain reference. I will dig on that anyways, but it will take me a couple of days more maybe